backends/backend_openssl3.c

```diff
@@ -2472,3 +2472,620 @@ static void openssl_ecdsa_backend(void)
+}
+
+static int rsa_create_pkey_crt(EVP_PKEY **pkey,
+						const unsigned char *n, size_t n_len,
+						const unsigned char *e, size_t e_len,
+						const unsigned char *p, size_t p_len,
+						const unsigned char *q, size_t q_len,
+						const unsigned char *dmp1, size_t dmp1_len,
+						const unsigned char *dmq1, size_t dmq1_len,
+						const unsigned char *iqmp, size_t iqmp_len,
+						BN_CTX *bn_ctx)
+{
+	int ret = 0;
+	EVP_PKEY_CTX *ctx = NULL;
+	OSSL_PARAM_BLD *bld = NULL;
+	OSSL_PARAM *params = NULL;
+	BIGNUM *e_bn = NULL, *n_bn = NULL, *p_bn= NULL, *q_bn= NULL, *dmp1_bn= NULL, *dmq1_bn= NULL, *iqmp_bn= NULL;
+	bld = OSSL_PARAM_BLD_new();
+	n_bn = BN_CTX_get(bn_ctx);
+	BN_bin2bn(n, n_len, n_bn);
+	OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_N, n_bn);
+
+	if (e != NULL) {
+		e_bn = BN_CTX_get(bn_ctx);
+		BN_bin2bn(e, e_len, e_bn);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_E, e_bn);
+	}
+	if (p != NULL) {
+		p_bn = BN_CTX_get(bn_ctx);
+		BN_bin2bn(p, p_len, p_bn);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_FACTOR1, p_bn);
+	}
+	if (q != NULL) {
+		q_bn = BN_CTX_get(bn_ctx);
+		BN_bin2bn(q, q_len, q_bn);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_FACTOR2, q_bn);
+	}
+	if (dmp1 != NULL) {
+		dmp1_bn = BN_CTX_get(bn_ctx);
+		BN_bin2bn(dmp1, dmp1_len, dmp1_bn);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_EXPONENT1, dmp1_bn);
+	}
+	if (dmq1 != NULL) {
+		dmq1_bn = BN_CTX_get(bn_ctx);
+		BN_bin2bn(dmq1, dmq1_len, dmq1_bn);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_EXPONENT2, dmq1_bn);
+	}
+	if (iqmp != NULL) {
+		iqmp_bn = BN_CTX_get(bn_ctx);
+		BN_bin2bn(iqmp, iqmp_len, iqmp_bn);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_COEFFICIENT1, iqmp_bn);
+	}
+	if(p != NULL && q != NULL){
+		BIGNUM *d_bn = BN_dup(n_bn);
+		BN_sub(d_bn, d_bn, p_bn);
+		BN_sub(d_bn, d_bn, q_bn);
+		BN_add_word(d_bn, 1);
+		BN_mod_inverse(d_bn, e_bn, d_bn, bn_ctx);
+		OSSL_PARAM_BLD_push_BN(bld, OSSL_PKEY_PARAM_RSA_D, d_bn);
+	}
+
+	params = OSSL_PARAM_BLD_to_param(bld);
+	ctx = EVP_PKEY_CTX_new_from_name(NULL, "RSA", "fips=yes");
+	EVP_PKEY_fromdata_init(ctx);
+	ret = EVP_PKEY_fromdata(ctx, pkey, EVP_PKEY_KEYPAIR, params);
+	if(ret == 0)
+		logger(LOGGER_DEBUG, "Keygen failed\n");
+
+	return ret;
+}
+
+/************************************************
+ * CMAC/HMAC cipher interface functions
+ ************************************************/
+@@ -2569,3 +2692,498 @@ static void openssl_ecdsa_backend(void)
+{
+	register_ecdsa_impl(&openssl_ecdsa);
+}
+
+
+/************************************************
+ * RSA interface functions
+ ************************************************/
+
+#define RSA 		EVP_PKEY
+#define RSA_free(a)		EVP_PKEY_free(a)
+#define EVP_DigestSignInit(a,b,c,d,e)		EVP_DigestSignInit_ex(a,b,EVP_MD_name(c),NULL,NULL,e,NULL)
+#define EVP_DigestVerifyInit(a,b,c,d,e)		EVP_DigestVerifyInit_ex(a,b,EVP_MD_name(c),NULL,NULL,e,NULL)
+
+static int openssl_rsa_keygen_prime(struct rsa_keygen_prime_data *data,
```

Comment:
According to the OpenSSL source code (https://github.com/openssl/openssl/blob/master/crypto/rsa/rsa_sp800_56b_gen.c), they use Appendix B.3.6 to generate primes. rsa_keygen_prime is used for B.3.3.